File: 0079-word-search/0079-word-search.java
class Solution {
    char[][] b;
    int[][] dir = {{-1,0},{1,0},{0,-1},{0,1}};
    boolean rs = false;
    
    public boolean exist(char[][] board, String word) {
        b = board;
        for(int x=0;x<b.length;x++){
            if(rs) break;
            for(int y=0;y<b[0].length;y++){
                if(b[x][y] == word.charAt(0) && backTrack(x,y,word)) {
                    rs = true;
                    break;
                }
            }
        }
        
        return rs;
    }
    
    public boolean backTrack(int x, int y, String w) {
        if(x < 0 || x >= b.length ||
                y < 0 || y >= b[0].length) return false;

        if(b[x][y] != w.charAt(0)) return false;
        if(w.length()==1) return true;

        b[x][y] = '.';
        for(int i=0;i<dir.length;i++){
            if(backTrack(x+dir[i][0],y+dir[i][1],w.substring(1))) return true;
        }

        b[x][y] = w.charAt(0);
        return false;
    }
}

File: 0079-word-search/0079-word-search.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        rows , cols = len(board), len(board[0])
        tot_cells = rows * cols
        if len(word) > tot_cells: return False

        board_cnt = Counter()
        for row in board:
            board_cnt.update(row)
        word_cnt = Counter(word)

        for c in word:
            if word_cnt[c] > board_cnt[c]:
                return False

        path = set()
        def dfs(r, c, i):
            if i == len(word):
                return True
            
            if (min(r, c) < 0 or
                r >= rows or c >= cols or
                board[r][c] != word[i] or
                (r, c) in path):
                return False
            
            path.add((r, c))
            res = (dfs(r, c + 1, i + 1) or
                    dfs(r + 1, c, i + 1) or
                    dfs(r, c - 1, i + 1) or
                    dfs(r - 1, c, i + 1))
            path.remove((r, c))
            return res

        for i in range(rows):
            for j in range(cols):
                if dfs(i, j, 0):
                    return True
        return False